drivers/adc/ad400x/iio_ad400x.c

```diff
@@ -0,0 +1,243 @@
+#include <stdio.h>
+#include <math.h>
+#include <string.h>
+
+#include "ad400x.h"
+#include "iio_ad400x.h"
+#include "iio.h"
+#include "no_os_util.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+
+static struct scan_type ad400x_iio_scan_types[] = {
+	[ID_AD4000] = { .sign = 'u', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4001] = { .sign = 's', .realbits = 16, .storagebits = 32, .is_big_endian = true, .shift = 16},
+	[ID_AD4002] = { .sign = 'u', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4003] = { .sign = 's', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4004] = { .sign = 'u', .realbits = 16, .storagebits = 32, .is_big_endian = true, .shift = 16},
+	[ID_AD4005] = { .sign = 's', .realbits = 16, .storagebits = 32, .is_big_endian = true, .shift = 16},
+	[ID_AD4006] = { .sign = 'u', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4007] = { .sign = 's', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4011] = { .sign = 's', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+	[ID_AD4020] = { .sign = 's', .realbits = 20, .storagebits = 32, .is_big_endian = true, .shift = 12},
+	[ID_ADAQ4003] = { .sign = 's', .realbits = 18, .storagebits = 32, .is_big_endian = true, .shift = 14},
+};
+
+/**
+ * @brief Read the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to read data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad400x_debug_reg_read(void *device, uint32_t reg,
+		uint32_t *val)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	int ret;
+	uint8_t val8;
+
+	ret = ad400x_spi_reg_read(dev, &val8);
+	if (ret)
+		return ret;
+
+	*val = val8;
+	return 0;
+}
+
+/**
+ * @brief Write the debug register value
+ * @param device - Pointer to IIO device instance
+ * @param reg - Register address to read from
+ * @param val - Pointer to variable to write data into
+ * @return 0 in case of success, negative value otherwise
+ */
+static int32_t iio_ad400x_debug_reg_write(void *device, uint32_t reg,
+		uint32_t val)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+
+	return ad400x_spi_reg_write(dev, (uint8_t)val);
+}
+
+static int get_raw(void *device, char *buf, uint32_t len,
+		   const struct iio_ch_info *channel, intptr_t id)
+{
+	struct ad400x_iio_dev *iio_dev = device;
+	struct ad400x_dev *dev = iio_dev->ad400x_dev;
+	struct scan_type *scan_type = &ad400x_iio_scan_types[dev->dev_id];
+	uint32_t results;
+	int ret;
+
+	ret = ad400x_spi_single_conversion(dev, &results);
+	if (ret)
+		return ret;
+
+	results = no_os_get_unaligned_be32((uint8_t *)&results);
```

Comment:
actually, this is not entirely correct. so the ad400x_spi_single_conversion is expected to return the conversion result as read from the device but in the processor endianess. so this no_os_get_unaligned_be32 should actually be called in ad400x_spi_single_conversion, after reading the SPI data... further more it should also be shifted (in my opinion).

Reply:
is it not up to the final application to check the scan_type of iio_info, and apply all these bit operations to display the raw_data correctly?

i would think that the driver would want to get sample data out as fast as possible. if we do bit operations in ad400x_spi_single_conversion it will affect also the loop in iio_ad400x_submit_buffer.. which will have a performance impact.

Reply:
@rbolboac are you ok with leaving this bit operations in this function rather than moving them to ad400x_spi_single_conversion? 

if i do move the bit operation i would have to:

- wrap the bit operations in #if flags for simple spi as HDL does not return the same byte/bit order.
- crate a new function that does not apply bit operation for use in the get data buffer loop

Reply:
It is fine for me but please describe in ad400x_spi_single_conversion API the exact format of the result

Reply:
ill double check. (need to build HDL for 400x) but from what i see in the code the exact format of the result seems to depend if we use HDL or simple spi. I think HDL might be doing some magic manipulations.

Reply:
please keep in mind that the result should be the same for both platforms